dali/core/exec/tasking_test.cc

```diff
@@ -0,0 +1,374 @@
+    return t->GetInputValue<InstanceCounter<int>>(0).payload;
+  });
+  // This task consumes the 1st output of producer
+  consumer1->Subscribe(producer, 0)->Succeed(sem1);
+
+  auto consumer2 = Task::Create([](Task *t) {
+    return t->GetInputValue<InstanceCounter<int>>(0).payload;
+  });
+  // This task consumes the 2nd output of producer
+  consumer2->Subscribe(producer, 1)->Succeed(sem2);
+
+  auto consumer3 = Task::Create([](Task *t) {
+    return t->GetInputValue<InstanceCounter<int>>(0).payload;
+  });
+  // This task consumes the (again) 2nd output of producer
+  consumer3->Subscribe(producer, 1)->Succeed(sem3);
+
+
+  ex.AddSilentTask(producer);
+  ex.AddSilentTask(consumer1);
+  ex.AddSilentTask(consumer2);
+  ex.AddSilentTask(consumer3);
+  ex.Wait(producer);
+
+  // Once producer is finished we should still see both instances - they should be in the inputs
+  // of the consumers.
+  EXPECT_EQ(InstanceCounter<int>::num_instances, 2);
+
+  // We trigger 1st consumer
+  sem1->Release(ex);
+  ex.Wait(consumer1);
+  // After it's done, the 1st output of producer is no longer needed - it should be destroyed
+  EXPECT_EQ(InstanceCounter<int>::num_instances, 1);
+  sem2->Release(ex);
+  ex.Wait(consumer2);
+  // 2nd output is different - it has 2 consumers
+  EXPECT_EQ(InstanceCounter<int>::num_instances, 1);
+
+  sem3->Release(ex);
+  ex.Wait(consumer3);
+  // Both consumers are gone - the 2nd output should be gone now.
+  EXPECT_EQ(InstanceCounter<int>::num_instances, 0);
+}
+
+namespace {
+
+template <typename RNG>
+std::vector<int> subset(int n, int m, RNG &rng) {
+  std::vector<int> selected;
+  for (int i = 0; i < m; i++) {
+    std::uniform_int_distribution<> dist(0, n - i - 1);
+    int k = dist(rng);
+    int j = 0;
+    for (; j < static_cast<int>(selected.size()); j++) {
+      if (selected[j] <= k)
+        k++;
+      else
+        break;
+    }
+    selected.insert(selected.begin() + j, k);
+  }
+  return selected;
+}
+
+double slowfunc(double x) {
+  return std::sqrt(x + std::sqrt(x) + std::sqrt(x + std::sqrt(x)));
+}
+
+// This test creates a large graph with several layers and randomly placed connections
+// between nodes in the layers.
+// Each node produces a value that is then consumed by the dependent tasks.
+// Each task compares the value vs a reference computed at task creation.
+// In the event of failure, obtaining the future throws an exception.
+void GraphTest(Executor &ex, int num_layers, int layer_size, int prev_layer_conn) {
+  std::vector<SharedTask> tasks;
+  std::set<Task *> has_successor;
+  std::mt19937_64 rng;
+  tasks.reserve(num_layers * layer_size);
+  std::vector<int> values(num_layers * layer_size);
+  std::uniform_int_distribution<> vdist(0, 10000);
+  for (auto &v : values)
+    v = vdist(rng);
+  for (int l = 0; l < num_layers; l++) {
+    int prev_layer_start = l ? tasks.size() - layer_size : 0;
+    int layer_start = tasks.size();
+    int prev_layer_size = layer_start - prev_layer_start;
```

Comment:
```suggestion
    int prev_layer_size = layer_size;
```
?

Reply:
The 1st layer doesn't have a previous layer.

Reply:
So it is either 0 or layer_size?